File: android/app/src/main/kotlin/pt/ualg/miaugenda/data/remote/UserApi.kt
package pt.ualg.miaugenda.data.remote

import pt.ualg.miaugenda.data.model.CreateUserRequest
import pt.ualg.miaugenda.data.model.UpdateProfileRequest
import pt.ualg.miaugenda.data.model.User
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path

interface UserApi {

    @GET("users")
    suspend fun getUsers(): Response<List<User>>

    @GET("users/me")
    suspend fun getMe(): Response<User>

    @PATCH("users/me")
    suspend fun updateMe(
        @Body request: UpdateProfileRequest
    ): Response<User>

    @POST("users")
    suspend fun createUser(
        @Body request: CreateUserRequest
    ): Response<User>

    @PATCH("users/{id}/deactivate")
    suspend fun deactivateUser(
        @Path("id") id: Int
    ): Response<User>
}

File: desktop/src/api/userApi.ts
import { apiClient } from "./apiClient";
import type {
  CreateUserRequest,
  UpdateProfileRequest,
  UpdateUserRequest,
  User
} from "../models/user";

export async function getUsers(): Promise<User[]> {
  const response = await apiClient.get<User[]>("/users");
  return response.data;
}

export async function getMe(): Promise<User> {
  const response = await apiClient.get<User>("/users/me");
  return response.data;
}

export async function updateMe(request: UpdateProfileRequest): Promise<User> {
  const response = await apiClient.patch<User>("/users/me", request);
  return response.data;
}

export async function createUser(request: CreateUserRequest): Promise<User> {
  const response = await apiClient.post<User>("/users", request);
  return response.data;
}

export async function updateUser(id: number, request: UpdateUserRequest): Promise<User> {
  const response = await apiClient.patch<User>(`/users/${id}`, request);
  return response.data;
}

export async function deactivateUser(id: number): Promise<User> {
  const response = await apiClient.patch<User>(`/users/${id}/deactivate`);
  return response.data;
}

export async function activateUser(id: number): Promise<User> {
  const response = await apiClient.patch<User>(`/users/${id}/activate`);
  return response.data;
}

export async function uploadAvatar(file: File): Promise<User> {
  const form = new FormData();
  form.append("avatar", file);
  const response = await apiClient.patch<User>("/users/me/avatar", form, {
    headers: { "Content-Type": "multipart/form-data" }
  });
  return response.data;
}